New flow

diegoquinteiro opened the pull request:
This PR replaces the existing settings screen with a new layout and flow.

See screenshots:
![1](https://cloud.githubusercontent.com/assets/1878108/17560904/2c36b4de-5efa-11e6-82e9-9b0158e05ebf.png)
![2](https://cloud.githubusercontent.com/assets/1878108/17560902/2c34306a-5efa-11e6-8f9f-ea54b54504ca.png)
![3](https://cloud.githubusercontent.com/assets/1878108/17560903/2c35efd6-5efa-11e6-8a32-3d27736272cb.png)
![4](https://cloud.githubusercontent.com/assets/1878108/17560901/2c33c97c-5efa-11e6-9489-79ff6feccbe5.png)
![5](https://cloud.githubusercontent.com/assets/1878108/17560899/2c308780-5efa-11e6-8574-fa7ef611c39a.png)
![6](https://cloud.githubusercontent.com/assets/1878108/17560900/2c31606a-5efa-11e6-8c61-2d1f8c59dec1.png)

diegoquinteiro commented:
Hey @philipjohn, thanks a lot for looking at this in a short timeline!

Here's the WIP PR you can comment on. I'll be finishing the last feature today and address any feedback you provide before the EOW.

diegoquinteiro commented:
@philipjohn we're ready for review =)

philipjohn commented on facebook-instant-articles.php:
```diff
@@ -89,6 +89,23 @@ function instant_articles_activate() {
 	}
 	register_activation_hook( __FILE__, 'instant_articles_activate' );
 
+	add_action( 'admin_init', 'instant_articles_redirect_settings' );
```
`instant_articles_redirect_settings()` doesn't seem to exist :)

philipjohn commented on wizard/class-instant-articles-option.php:
```diff
@@ -454,4 +454,37 @@ public function universal_sanitize_and_encode_handler( $payload ) {
+			$option_key = static::OPTION_KEY;
+		}
+
+		wp_cache_delete ( 'alloptions', 'options' );
```
These seems a bit excessive - can we do a more targeted delete operation here?

philipjohn replied:
On reflection, I wonder if this is even needed? We should trust WordPress to handle the "update_option" and delete the relevant caches itself, rather than forcing deletion of a bunch of stuff ourselves.

diegoquinteiro replied:
That was done because update_option was failing when called on AJAX. According to:

https://developer.wordpress.org/reference/functions/update_option/

On the second to last comment, that's the workaround for when the cache is not being invalidated correctly.

philipjohn commented on wizard/class-instant-articles-option.php:
```diff
@@ -454,4 +454,37 @@ public function universal_sanitize_and_encode_handler( $payload ) {
+			$option_key = static::OPTION_KEY;
+		}
+
+		wp_cache_delete ( 'alloptions', 'options' );
```
See line 471 comment

philipjohn commented on wizard/class-instant-articles-wizard-review-submission.php:
```diff
@@ -0,0 +1,93 @@
+
+	public static function getArticlesForReview() {
+		$post_types = apply_filters( 'instant_articles_post_types', array( 'post' ) );
+		return wp_get_recent_posts(
```
`wp_get_recent_posts()` uses `get_posts()` which is an uncached function. `WP_Query` should be used instead here.

diegoquinteiro replied:
It looks like get_posts() simply uses WP_Query now. Can you double check?

https://core.trac.wordpress.org/browser/tags/4.5.3/src/wp-includes/post.php#L1735

philipjohn commented on wizard/class-instant-articles-wizard-review-submission.php:
```diff
@@ -0,0 +1,93 @@
+		$post_types = apply_filters( 'instant_articles_post_types', array( 'post' ) );
+		return wp_get_recent_posts(
+		 	array(
+				'numberposts' => self::MIN_ARTICLES,
```
Let's use `min()` here to put a hard-limit of 100 on this

diegoquinteiro replied:
Good call (y)

philipjohn commented on wizard/class-instant-articles-wizard-state.php:
```diff
@@ -0,0 +1,352 @@
+	public static function get_current_state() {
+
+		$option = get_option( 'instant-articles-current-state', null );
+		if ( in_array( $option, array_keys( self::$transition_vectors ) ) ) {
```
`$option` might not be an array, which would make this produce an error. We should do an `is_array()` check before the `in_array()`

diegoquinteiro replied:
Ops! Actually `$option` is the needle not the haystack, so it does not need to be an array.

philipjohn commented on wizard/class-instant-articles-wizard-state.php:
```diff
@@ -0,0 +1,352 @@
+		} catch (Exception $e) {
+			// Here we override the error message to give an actionable
+			// instruction to the user that is specific for WordPress.
+			throw new Exception("Could not claim the URL for the selected page, please make sure your site is publicly available.");
```
We need to avoid fatal errors anywhere, so we should deal with this situation in some other way than throwing an exception.

diegoquinteiro replied:
We do capture this exception on the call site, so it never leads to a fatal error.

philipjohn commented on wizard/class-instant-articles-wizard.php:
```diff
@@ -0,0 +1,291 @@
+	public static function add_settings_link_to_plugin_actions( $links ) {
+		$link_text = __( 'Settings' );
+		$settings_href = self::get_url();
+		$settings_link = '<a href="' . esc_url( $settings_href ) . '">' . $link_text . '</a>';
```
`$link_text` should be wrapped in `esc_html()` here

diegoquinteiro replied:
👍

philipjohn commented on wizard/class-instant-articles-wizard.php:
```diff
@@ -0,0 +1,291 @@
+			wp_die( esc_html( 'You do not have sufficient permissions to access this page.' ) );
+		}
+
+		$new_state = filter_input( INPUT_POST, 'new_state' );
```
We should use WordPress' own sanitisation functions here with `$_POST['new_state']` rather than `filter_input`.

philipjohn commented on wizard/class-instant-articles-wizard.php:
```diff
@@ -0,0 +1,291 @@
+		}
+
+		$new_state = filter_input( INPUT_POST, 'new_state' );
+		$params = filter_input( INPUT_POST, 'params' );
```
As line 112

philipjohn commented on wizard/class-instant-articles-wizard.php:
```diff
@@ -0,0 +1,291 @@
+			die();
+		}
+
+		$app_id = filter_input( INPUT_POST, 'app_id' );
```
As 112 - let's use WP sanitisation

philipjohn commented on wizard/class-instant-articles-wizard.php:
```diff
@@ -0,0 +1,291 @@
+		$pages = $fb_helper->get_pages();
+
+		if ( isset( $pages[ $page_id ] ) && $pages[ $page_id ][ 'supports_instant_articles' ] ) {
+			die( 'yes' );
```
This could probably be done better, but works for now

diegoquinteiro replied:
It looks very hacky indeed. Do you have any specific recommendation?

philipjohn commented on wizard/templates/app-setup-template.php:
```diff
@@ -0,0 +1,115 @@
+					<li>Create an app name, input your email and select 'Apps for Pages' in the 'Category' dropdown menu. Click 'Create App ID' when you’re ready.</li>
+					<li>Click on 'Settings' in the left nav bar.</li>
+					<li>Click '+Add Platform' at the bottom of the page and select 'Website.'</li>
+					<li>Under both Website and App Domains enter your domain: <b><?php echo self::get_admin_url(); ?></b></li>
```
We should use `esc_url()` here

diegoquinteiro replied:
👍

philipjohn commented on wizard/templates/overview-template.php:
```diff
@@ -0,0 +1,43 @@
+			<h3>Get Started with the Instant Articles WordPress Plugin</h3>
+			<div class="instant-articles-card-steps">
+				<div class="instant-articles-card-step">
+					<img src="<?php echo esc_attr( plugin_dir_url( __FILE__ ) ); ?>../../assets/key@2x.png">
```
`esc_url()` should be used for escaping URLs over `esc_attr()`

philipjohn commented on wizard/templates/overview-template.php:
```diff
@@ -0,0 +1,43 @@
+					<p>If you don't have one already, set up your Facebook Developers App. Then log in to connect your plugin.</p>
+				</div>
+				<div class="instant-articles-card-step">
+					<img src="<?php echo esc_attr( plugin_dir_url( __FILE__ ) ); ?>../../assets/connect@2x.png">
```
`esc_url()` should be used for escaping URLs over `esc_attr()`

philipjohn commented on wizard/templates/overview-template.php:
```diff
@@ -0,0 +1,43 @@
+					<p>Select the Page you'll use to publish your Instant Articles.</p>
+				</div>
+				<div class="instant-articles-card-step">
+					<img src="<?php echo esc_attr( plugin_dir_url( __FILE__ ) ); ?>../../assets/customize@2x.png">
```
`esc_url()` should be used for escaping URLs over `esc_attr()`

philipjohn commented on wizard/templates/overview-template.php:
```diff
@@ -0,0 +1,43 @@
+					<p>Use our Style Editor to make your Instant Articles look just how you want them to.</p>
+				</div>
+				<div class="instant-articles-card-step">
+					<img src="<?php echo esc_attr( plugin_dir_url( __FILE__ ) ); ?>../../assets/check@2x.png">
```
`esc_url()` should be used for escaping URLs over `esc_attr()`

philipjohn commented on wizard/templates/page-selection-template.php:
```diff
@@ -0,0 +1,63 @@
+			<span class="instant-articles-card-title-checkmark">✔</span>
+			<label class="instant-articles-card-title-label">App connected:</label>
+			<span class="instant-articles-card-title-value"><?php echo esc_html( $fb_app_settings[ 'app_id' ] ); ?></span>
+			<a class="instant-articles-wizard-transition instant-articles-card-title-edit" href="#" data-new-state="<?php echo esc_html( Instant_Articles_Wizard_State::STATE_APP_SETUP ); ?>"></a>
```
`esc_attr()` rather than `esc_html()` should be used here

philipjohn commented on wizard/templates/page-selection-template.php:
```diff
@@ -0,0 +1,63 @@
+							<?php else : ?>
+								<span class="page-not-enabled">
+									This page has not been signed up yet.
+									<a href="https://web.facebook.com/instant_articles/signup?redirect_uri=<?php echo urlencode( $settings_url ) ?>&page_id=<?php echo urlencode( $page[ 'page_id' ] ) ?>">Sign Up</a>.
```
For this URL output, either the individual vars should be escaped with `rawurlencode()` or the whole lot with `esc_url()`

philipjohn commented on wizard/templates/review-submission-template.php:
```diff
@@ -0,0 +1,156 @@
+			<span class="instant-articles-card-title-checkmark">✔</span>
+			<label class="instant-articles-card-title-label">App connected:</label>
+			<span class="instant-articles-card-title-value"><?php echo esc_html( $fb_app_settings[ 'app_id' ] ); ?></span>
+			<a class="instant-articles-wizard-transition instant-articles-card-title-edit" href="#" data-new-state="<?php echo esc_html( Instant_Articles_Wizard_State::STATE_APP_SETUP ); ?>"></a>
```
`esc_attr()` should be used here over `esc_html()`

philipjohn commented on wizard/templates/review-submission-template.php:
```diff
@@ -0,0 +1,156 @@
+			<?php if ( $fb_page_settings[ 'page_picture' ] ) : ?>
+				<img class="instant-articles-page-img" src="<?php echo esc_attr( $fb_page_settings[ 'page_picture' ] ) ?>"/>
+			<?php endif; ?>
+			<a class="instant-articles-wizard-transition instant-articles-card-title-edit" href="#" data-new-state="<?php echo esc_html( Instant_Articles_Wizard_State::STATE_PAGE_SELECTION ); ?>"></a>
```
`esc_attr()` should be used here over `esc_html()`

philipjohn commented on wizard/templates/review-submission-template.php:
```diff
@@ -0,0 +1,156 @@
+	<div class="instant-articles-card-title">
+		<h3>Style Customized</h3>
+		<div class="instant-articles-card-title-right">
+			<a class="instant-articles-wizard-transition instant-articles-card-title-edit" href="#" data-new-state="<?php echo esc_html( Instant_Articles_Wizard_State::STATE_STYLE_SELECTION ); ?>"></a>
```
`esc_attr()` should be used here over `esc_html()`

philipjohn commented on wizard/templates/review-submission-template.php:
```diff
@@ -0,0 +1,156 @@
+					</p>
+					<p>
+						Next, set up monetization and analytics in Advanced Settings. Or explore the
+						<a href="https://www.facebook.com/<?php echo esc_attr( $fb_page_settings['page_id'] ); ?>/publishing_tools/?section=INSTANT_ARTICLES_SETTINGS" target="_blank">Instant Articles publishing tools</a>
```
`esc_url()` should be used for the whole URL here over `esc_attr()` for the variable

philipjohn commented on wizard/templates/review-submission-template.php:
```diff
@@ -0,0 +1,156 @@
+					</p>
+					<p>
+						Please go to your selected
+						<a href="https://www.facebook.com/<?php echo esc_attr( $fb_page_settings['page_id'] ); ?>/publishing_tools/?section=INSTANT_ARTICLES_SETTINGS#Step-2" target="_blank">Facebook Page's Publishing Tools</a>
```
`esc_url()` should be used for the whole URL here over `esc_attr()` for the variable

philipjohn commented on wizard/templates/style-selection-template.php:
```diff
@@ -0,0 +1,66 @@
+			<span class="instant-articles-card-title-checkmark">✔</span>
+			<label class="instant-articles-card-title-label">App connected:</label>
+			<span class="instant-articles-card-title-value"><?php echo esc_html( $fb_app_settings[ 'app_id' ] ); ?></span>
+			<a class="instant-articles-wizard-transition instant-articles-card-title-edit" href="#" data-new-state="<?php echo esc_html( Instant_Articles_Wizard_State::STATE_APP_SETUP ); ?>"></a>
```
`esc_attr()` should be used here rather than `esc_html()`

philipjohn commented on wizard/templates/style-selection-template.php:
```diff
@@ -0,0 +1,66 @@
+		<div class="instant-articles-card-title-right">
+			<a href="" class="instant-articles-card-title-link"><?php echo esc_html( $fb_page_settings[ 'page_name' ] ) ?></a>
+			<?php if ( $fb_page_settings[ 'page_picture' ] ) : ?>
+				<img class="instant-articles-page-img" src="<?php echo esc_attr( $fb_page_settings[ 'page_picture' ] ) ?>"/>
```
`esc_attr()` should be used here rather than `esc_attr()`

philipjohn commented on wizard/templates/style-selection-template.php:
```diff
@@ -0,0 +1,66 @@
+			<?php if ( $fb_page_settings[ 'page_picture' ] ) : ?>
+				<img class="instant-articles-page-img" src="<?php echo esc_attr( $fb_page_settings[ 'page_picture' ] ) ?>"/>
+			<?php endif; ?>
+			<a class="instant-articles-wizard-transition instant-articles-card-title-edit" href="#" data-new-state="<?php echo esc_html( Instant_Articles_Wizard_State::STATE_PAGE_SELECTION ); ?>"></a>
```
`esc_url()` should be used here rather than `esc_html()`

philipjohn commented on wizard/templates/style-selection-template.php:
```diff
@@ -0,0 +1,66 @@
+			</p>
+			<a
+				id="instant-articles-wizard-customize-style"
+				href="https://www.facebook.com/<?php echo esc_attr( $fb_page_settings['page_id'] ); ?>/publishing_tools/?section=INSTANT_ARTICLES_SETTINGS&style=<?php echo esc_attr($article_style); ?>"
```
`esc_url()` should be used here for the whole URL

philipjohn commented on wizard/templates/style-selection-template.php:
```diff
@@ -0,0 +1,66 @@
+				target="_blank">
+				Customize
+			</a>
+			<button id="instant-articles-wizard-customize-style-next" class="instant-articles-button instant-articles-wizard-transition" data-new-state="<?php echo esc_html( Instant_Articles_Wizard_State::STATE_REVIEW_SUBMISSION ); ?>">
```
`esc_attr()` should be used here rather than `esc_html()`

philipjohn commented:
Thanks @diegoquinteiro - I've left a bunch of inline comments. Mostly escaping choices that should be a quick fix. Otherwise looks good!

diegoquinteiro commented:
@philipjohn I've went ahead and merged after addressing the comments. Please let me know if you have new comments so we can address them in a new PR. Thanks!
